### PR TITLE
pkg/apiserver/diagnose/query.go: eliminate the gorm.DB.Raw()

### PR DESCRIPTION
**Pull request by zzh-wisdom**
UCP #108

## What have you changed? 

eliminate the `gorm.DB.Raw()`, and use `rows()`
Define a new function `scanRows`, which is used to scan `sql.Rows`
delete func `genSummarySQLs` and `genDetailSQLs`
add func `querySummary` and `queryDetail`

## What are the type of the changes? 

Improvement

## How has this PR been tested? 

Have run command make run successfully

## Does this PR affect documentation (docs/docs-cn) update?

No

### CODE REVIEW

**Comment by zzh-wisdom** [comment]
@baurine

**Comment by breeswish** [comment]
Hi, please avoid constructing SQLs that includes user specified parameters directly. Instead you need to use placeholders as examples in http://gorm.io/docs/sql_builder.html#Run-Raw-SQL and https://github.com/pingcap-incubator/tidb-dashboard/pull/34/commits/175c39a5f36b11388896a3acc09e53d2ffca0ea9 .

**Comment by breeswish** [comment]
Additionally the output of the SQL should be changed to schema aware, i.e. the result is an array of a structure, instead of HashMap.

**Comment by zzh-wisdom** [comment]
I don't quite understand what “use placeholders” means。
I guess I shoud change the codes such as
`conditions := fmt.Sprintf("t0.time >= '%s' and t0.time < '%s'", startTime, endTime)` 
into
`db.where("t0.time >= ? and t0.time < ?", startTime, endTime)`
is it?

**Comment by zzh-wisdom** [comment]
In addition to the above, do I need to modify other codes? 
For example,
`fields := fmt.Sprintf("'%s','', sum(t0.value)", t.name)`
`tables := fmt.Sprintf("metrics_schema.%s t0", t.sumTbl)`
the select fields and table name are also created by constructing

**Comment by zzh-wisdom** [comment]
@breeswish

**Comment by zzh-wisdom** [comment]
Some queries return non-fixed columns, how to use an array of structure to get the results.
For example,
`sql = fmt.Sprintf("select '%[1]v',`%[2]v`, sum(value) from metrics_schema.%[3]v %[4]s group by `%[2]v` having sum(value) > 0 order by sum(value) desc",
		t.name, strings.Join(t.labels, "`,`"), t.tbl, condition)`
it returns Variable `t.labels`.
@breeswish

**Comment by zzh-wisdom** [comment]
Hi, I have improved the code. Please review, thanks

**Comment by zzh-wisdom** [comment]
@breeswish

**Comment by zzh-wisdom** [comment]
@baurine

File: pkg/apiserver/diagnose/query.go [hunk 1]
```diff
@@ -55,30 +55,69 @@ func (t AvgMaxMinTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, e
...
+	row := avgMaxMinToSlice(result)
 	if len(t.label) == 0 {
-		return t.genRow(rows[0], nil), nil
+		return t.genRow(row, nil), nil
+	}
+
+	label := "`" + t.label + "`"
+	fields = fmt.Sprintf(`
+		'%s' as name,
+		%s as label,
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: Maybe this is better:

```
`%s` as label,
```

File: pkg/apiserver/diagnose/query.go [hunk 2]
```diff
@@ -382,67 +524,154 @@ func (t totalValueAndTotalCountTableDef) genRow(values []string, subValues [][]s
...
-		fields += fmt.Sprintf("t%v.*", i)
-		tbls += fmt.Sprintf(" (%s) as t%v ", sql, i)
+		aggrQuantileMaxField += fmt.Sprintf("max(%v.value)", tableAlias)
+		tableNum++
 	}
-	joinSQL := fmt.Sprintf("select %v from %v", fields, tbls)
-	return joinSQL
+	if len(aggrQuantileMaxField) == 0 {
+		aggrQuantileMaxField = "''"
+	}
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: you can avoid building the field list if `quantiles []float64` are changed to a closed set, like this:

```golang
type QuantileFields {
  Q999 float64
  Q99 float64
  Q90 float64
  Q80 float64
}

type Quantiles {
  Q999 bool
  Q99 bool
  Q90 bool
  Q80 bool
}
```

When using it:

```golang
type XxxRow {
  QuantileFields
}

func (...) querySummary(..., quantiles: Quantiles) ([]XxxRow, error) {
  ...
}
```

Actually quantiles are indeed a closed set:

```golang
quantiles: []float64{0.999, 0.99, 0.90, 0.80},
```

**Comment by zzh-wisdom** [comment]
> Hi, can you modify the return signature of these interfaces to make them structural? As you can see, finally all of these tables have a "schema":
> 
> ```go
> func GenerateReportError(errRows []TableRowDef) *TableDef {
> 	return &TableDef{
> 		Category:  []string{CategoryError},
> 		Title:     "Generate Report Error",
> 		CommentEN: "",
> 		CommentCN: "",
> 		Column:    []string{"CATEGORY", "TABLE", "ERROR"},
> 		Rows:      errRows,
> 	}
> }
> 
> func GetHeaderTimeTable(startTime, endTime string, db *gorm.DB) (TableDef, error) {
> 	return TableDef{
> 		Category:  []string{CategoryHeader},
> 		Title:     "Report Time Range",
> 		CommentEN: "",
> 		CommentCN: "",
> 		Column:    []string{"START_TIME", "END_TIME"},
> 		Rows: []TableRowDef{
> 			{Values: []string{startTime, endTime}},
> 		},
> 	}, nil
> }
> ```
> 
> So the best way is to simply make each table definition schema-aware (e.g. returning `[]AvgMaxMin`), instead of returning implicit schema data via `*TableRowDef` which is just an 2d array.

Does it mean I also need to modify the codes where use these interfaces?

**Comment by breeswish** [comment]
> > Hi, can you modify the return signature of these interfaces to make them structural? As you can see, finally all of these tables have a "schema":
> > ```go
> > func GenerateReportError(errRows []TableRowDef) *TableDef {
> > 	return &TableDef{
> > 		Category:  []string{CategoryError},
> > 		Title:     "Generate Report Error",
> > 		CommentEN: "",
> > 		CommentCN: "",
> > 		Column:    []string{"CATEGORY", "TABLE", "ERROR"},
> > 		Rows:      errRows,
> > 	}
> > }
> > 
> > func GetHeaderTimeTable(startTime, endTime string, db *gorm.DB) (TableDef, error) {
> > 	return TableDef{
> > 		Category:  []string{CategoryHeader},
> > 		Title:     "Report Time Range",
> > 		CommentEN: "",
> > 		CommentCN: "",
> > 		Column:    []string{"START_TIME", "END_TIME"},
> > 		Rows: []TableRowDef{
> > 			{Values: []string{startTime, endTime}},
> > 		},
> > 	}, nil
> > }
> > ```
> > 
> > 
> > So the best way is to simply make each table definition schema-aware (e.g. returning `[]AvgMaxMin`), instead of returning implicit schema data via `*TableRowDef` which is just an 2d array.
> 
> Does it mean I also need to modify the codes where use these interfaces?

Yes.

**Comment by zzh-wisdom** [comment]
I modified the code to my own understanding.
I found that in the code where calls function `func (t totalValueAndTotalCountTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error)`, the parameter `arg.quantiles` is either empty or four float64. So I added a field `flag bool` to the returned structure 
`type TotalValueAndTotalCount struct {
	Name       string
	Label      string
	TotalSum   float64
	TotalCount float64
	MaxValueOfQuantile
	flag bool
}`
Because wanting to avoid building quantile fields list, each time do query will always return four `max(value)` of each quantiles and sometimes the `arg.quantiles` , in the old code, is empty. In the new code, when the value of a field in structure `Quantile` is `0.0`, it means empty.

But in  `func (t totalTimeByLabelsTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error) `, I found that `arg.quantiles` was always not empty. So I did not add `flag bool` in
`
type TotalTimeByLabels struct {
	Name          string
	Label         string
	RatioSumValue float64
	TotalTime     float64
	TotalCount    float64
	MaxValueOfQuantile
}
`

I didn't modify the interface because it involves a lot of code. For each information query, I have defined the function separately and returned the structure type, such as ` []*TotalTimeByLabels`. Maybe this also is schema-aware.
@breeswish

**Comment by zzh-wisdom** [comment]
I modified the code to my own understanding.
I found that in the code where calls function `func (t totalValueAndTotalCountTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error)`, the parameter `arg.quantiles` is either empty or four float64. So I added a field `flag bool` to the returned structure 
```
type TotalValueAndTotalCount struct {
	Name       string
	Label      string
	TotalSum   float64
	TotalCount float64
	MaxValueOfQuantile
	flag bool
}
```
Because wanting to avoid building quantile fields list, each time do query will always return four `max(value)` of each quantiles and sometimes the `arg.quantiles` , in the old code, is empty. In the new code, when the value of a field in structure `Quantile` is `0.0`, it means empty.

But in  `func (t totalTimeByLabelsTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error) `, I found that `arg.quantiles` was always not empty. So I did not add `flag bool` in
```
type TotalTimeByLabels struct {
	Name          string
	Label         string
	RatioSumValue float64
	TotalTime     float64
	TotalCount    float64
	MaxValueOfQuantile
}
```

I didn't modify the interface because it involves a lot of code. For each information query, I have defined the function separately and returned the structure type, such as ` []*TotalTimeByLabels`. Maybe this also is schema-aware.
@breeswish

File: pkg/apiserver/diagnose/query.go [hunk 3]
```diff
@@ -55,53 +71,107 @@ func (t AvgMaxMinTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, e
...
+		min(value) as min
+		`, t.name)
+	table := fmt.Sprintf("metrics_schema.%s", t.tbl)
+
+	query := db.
+		Select(fields).
+		Table(table).
+		Where("time >= ? AND time < ?", arg.startTime, arg.endTime)
+	if len(t.condition) > 0 {
+		query = query.Where(t.condition)
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: I think this should be parameterized.

**Comment by zzh-wisdom** [comment]
I probably understand what you mean. I think it takes too many changes to achieve this goal. Many codes in the file `report.go` should be also modified. So I accept the second solution that opening another UCP to achieve the goal. I'm sorry that I can not modify the code properly. By the way, what else code I need to modify for this PR?

**Comment by zzh-wisdom** [comment]
@breeswish

**Comment by breeswish** [comment]
Hi, @zzh-wisdom For this PR, you can revert the change of the structural return if you wish to do it in another PR. This PR can focus on changing the raw SQL string using parameterized `where()` and other query primitives. Sorry for the inconvenience!

**Comment by zzh-wisdom** [comment]
@breeswish ,I have modified the codes. Please take a look.

File: pkg/apiserver/diagnose/query.go [hunk 4]
```diff
@@ -118,37 +152,65 @@ func (t sumValueQuery) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error
...
-	if len(rows) == 0 {
+	if len(results) == 0 {
 		return nil, nil
 	}
 	if len(t.labels) == 0 {
-		return t.genRow(rows[0], nil), nil
+		return t.genRow(results[0], nil), nil
+	}
+
+	labelFields := fmt.Sprintf("`%v`", strings.Join(t.labels, "`||','||`"))
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: the original SQL is assembled using `strings.Join(t.labels, "`,`")`

File: pkg/apiserver/diagnose/query.go [hunk 5]
```diff
@@ -118,37 +152,65 @@ func (t sumValueQuery) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error
...
+		Rows()
 	if err != nil {
 		return nil, err
 	}
-	for i := range subRows {
-		row := subRows[i]
-		row[1] = strings.Join(row[1:1+len(t.labels)], ",")
-		newRow := row[:2]
-		newRow = append(newRow, row[1+len(t.labels):]...)
-		subRows[i] = newRow
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: maybe need to revert?

File: pkg/apiserver/diagnose/query.go [hunk 6]
```diff
@@ -181,40 +243,26 @@ type totalTimeByLabelsTableDef struct {
...
 func (t totalTimeByLabelsTableDef) queryRow(arg *queryArg, db *gorm.DB) (*TableRowDef, error) {
-	sql := t.genSumarySQLs(arg.totalTime, arg.startTime, arg.endTime, arg.quantiles)
-	rows, err := querySQL(db, sql)
+	rows, err := t.querySummary(db, arg.totalTime, arg.startTime, arg.endTime, arg.quantiles)
 	if err != nil {
 		return nil, err
 	}
-	if len(rows) == 0 {
-		return nil, nil
-	}
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: why can we remove these lines?

File: pkg/apiserver/diagnose/query.go [hunk 7]
```diff
@@ -181,40 +243,26 @@ type totalTimeByLabelsTableDef struct {
...
 	}
-
-	for i := range subRows {
-		row := subRows[i]
-		row[1] = strings.Join(row[1:1+len(t.labels)], ",")
-		newRow := row[:2]
-		newRow = append(newRow, row[1+len(t.labels):]...)
-		subRows[i] = newRow
-	}
-	return t.genRow(rows[0], subRows), nil
```
Comment (breeswish) [inline pkg/apiserver/diagnose/query.go]: ditto

File: pkg/apiserver/diagnose/query.go (hunk 4 above)
Comment (zzh-wisdom) [reply pkg/apiserver/diagnose/query.go]: `||` is used as a connector in `gorm`, equivalent to concatenating strings

**Comment by zzh-wisdom** [comment]
@breeswish  Are you convenient to take a look

**Pull request closed without merging**